raft-tech/arcadedb
# 1358 arcade keycloak rbac

## What does this PR do?

1. Arcade authorizes requests based on the presence of arcade specific roles in keycloak for the user or any user groups. This permission evaluation is done at the data access layer, so it authorizes requests regardless of which interface or protocol the request comes through.
   a. There are 3 categories of roles: user, dba, and sa.
   b. Keycloak roles will have the format of:

         [arcade prefix]__[role type]__[permission args...]

         [arcade prefix]__sa__[sa role]
         [arcade prefix]__dba__d-[database]__[dba role]
         [arcade prefix]__user__d-[database]__t-[table regex]__p-[crud]__[optional additional args]

3. Arcade will auto create new user accounts in it's internal ACLs for new keycloak users as they log in.
4. When a user creates a database, arcade will auto create a new role granting full access to the database, and assign it to the user, so they can immediately start hitting the database
5. Flips the data access permission evaluation of Arcade from "permissive to restrictive". Got rid of the default Read access to everything if there wasn't a specific group assigned to an object for the user.

This is a first pass. Additional improvements (tickets pending) are:
1. Cache keycloak metadata for requests
2. Cache user permissions data, poll for updates
3. Fire kafka events for user/group and role changes
6. Update arcade cache based on events
7. Test coverage, ideally before updating our fork
8. refactor a good chunk of these changes into plugins, to reduce our change collisions when updating the fork
9. Redirect arcdae logins to the keycloak login page
10. Move arcade -> keycloak rest calls to an arcade specific client (ideally after the keycloak config service is done)
11. Move arcdae -> keyclaok rest calls to use an arcade specific service account (ideally after the keycloak config service is done)
12. Incorrect login credentials in the arcade GUI return a meaningless "response already started error", rather than "not authorized"
13. Regex support for database names roles
14. Regex support for table names in roles
15. Support for multiple DBA/SA permissions in a single role
16. ...

## Motivation
Want to centralize user ACL management in keycloak. Prevent doing dual/triple maintenance of user accounts across all of our tools

## Related issues
Fixes
https://github.com/raft-tech/data-fabric/issues/1358
https://github.com/raft-tech/data-fabric/issues/2000

## Additional Notes
Will require the corresponding data-fabric PR, https://github.com/raft-tech/data-fabric/pull/2235

## Testing
### Build & deploy
1. Switch to the related dafa-fabric PR branch https://github.com/raft-tech/data-fabric/pull/2235
17. Deploy data fabric

        ./hacks/df_create_kind.sh
        export GHP_USERNAME=your_github_username_goes_here
        export GHP_SECRET=your_github_pat_goes_here
        ./hacks/df_deploy.sh

18. From the project root, run 

        docker buildx build --platform linux/arm64 --load -t raft-tech/arcadedev ./
        
19. Load into kind 

        kind load docker-image raft-tech/arcadedev  --name data-fabric
        
20. Update image in arcadedb's stateful set
21. Kill the pod to force it to take the change

### Testing
1. Confirm you can log into arcade as the admin [http://localhost/api/v1/arcadedb/studio](http://localhost/api/v1/arcadedb/studio)
2. Create a new database through the gui (click the database icon in the upper left, and look at the menu in the upper right)
3. Confirm you can switch to the new database in the database dropdown in the upper middle left of the screen
5. Confirm you can import data to the database by running the following script

        import database file://demo-data/FoodData_Central_foundation_food_json_2022-10-28.json with mapping = {
        "FoundationFoods": [{
		"@cat": "v",
		"@type": "<foodClass>",
		"foodNutrients": [{
			"@cat": "e",
			"@type": "HAS_NUTRIENT",
			"@in": "nutrient",
			"@cardinality": "no-duplicates",
			"nutrient": {
				"@cat": "v",
				"@type": "Nutrient",
				"@id": "id",
				"@idType": "long",
				"@strategy": "merge"
			},
			"foodNutrientDerivation": "@ignore"
		}],
		"inputFoods": [{
			"@cat": "e",
			"@type": "INPUT",
			"@in": "inputFood",
			"@cardinality": "no-duplicates",
			"inputFood": {
				"@cat": "v",
				"@type": "<foodClass>",
				"@id": "fdcId",
				"@idType": "long",
				"@strategy": "merge",
				"foodCategory": {
					"@cat": "e",
					"@type": "HAS_CATEGORY",
					"@cardinality": "no-duplicates",
					"@in": {
						"@cat": "v",
						"@type": "FoodCategory",
						"@id": "id",
						"@idType": "long",
						"@strategy": "merge"
					        }
				        }
			        }
		        }]
	        }]
        }
6. Wait a minute for data to be indexed and available
7. Run the following queries and confirm it returns data

        select *, bothE() as `@edges` from `Composite` limit 30
        select *, bothE() as `@edges` from `FinalFood` limit 30

22. Create a new df-backend client roles in keycloak with the following names. Insert the name of the database you previously created: 

        arcade__user__d-[database name]__t-Composite__p-CRUD                     Composite: All (Create/Read/Update/Delete)
        arcade__user__d-[database name]__t-Composite__p-CR                          Composite: Create/Read
        arcade__user__d-[database name]__t-Composite__p-RU                          Composite: Update/Update (Update requires read)
        arcade__user__d-[database name]__t-Composite__p-RD                          Composite: Delete/Read (delete requires read)
        arcade__dba__d-[database name]__updateSchema                                    Manage db schema

23. Create a new user in keycloak.
    a. Assign a password
24. Assign just the following role: `arcade__user__d-[database name]__t-Composite__p-CRUD `
25. log into arcade as the new account. You can logout by clicking the last vertical menu button on the left.
    a. confirm you **can run** all of the below CRUD commands
    b. confirm you are **unable to run** all the below CRUD commands if you update the type/table to `FinalFood`
    c. confirm you are **unable to run** the update schema script
26. Go to keycloak, add the update schema role to your user
27. Logout, log back in to arcade. Confirm the update schema command works
28. In keycloak, remove the arcade roles from your user, and add `arcade__user__d-[database name]__t-Composite__p-CR`
29. Logout/login to arcade. Confirm you can run the create and read scripts. Confirm the update and delete fail
30. Go to keycloak, remove the arcade roles for the user, and add `arcade__user__d-[database name]__t-Composite__p-RU `
31. Logout/login to arcade. Confirm the Read and update scripts work. Confirm Create and Delete fail.
32. Go to keycloak, remove the arcade roles for the user, and add ` arcade__user__d-[database name]__t-Composite__p-RD  `
33. Logout/login to arcade. Confirm the read and delete commands work. Confirm create and update fail.

### Use the following commands to test permissions:
#### Create (run each line individually) 

        insert into Composite CONTENT {foodClass: "Composite", description: "abc"}
        SELECT * FROM Composite  WHERE description="abc"

#### Read  

        select from `Composite` limit 30

#### Update (run each line individually)  

        UPDATE Composite set alpha = 1 WHERE description="abc"
        SELECT * FROM Composite  WHERE description="abc"

#### Delete (run each line individually)

        DELETE FROM Composite WHERE description="abc"
        SELECT * FROM Composite  WHERE description="abc"

#### Update schema

        CREATE vertex type Sugar
